src/libraries/System.Private.CoreLib/src/System/Environment.SunOS.cs

```diff
@@ -7,7 +7,20 @@ namespace System
 {
     public static partial class Environment
     {
-        public static long WorkingSet =>
-            (long)(Interop.procfs.TryReadProcessStatusInfo(Interop.procfs.ProcPid.Self, out Interop.procfs.ProcessStatusInfo status) ? status.ResidentSetSize : 0);
+        public static long WorkingSet
+        {
+            get
+            {
+                Interop.procfs.ProcessInfo iProcInfo;
+                if (Interop.procfs.GetProcessInfoById(ProcessId, out iProcInfo))
+                {
+                    return (long)iProcInfo.ResidentSetSize;
+                }
+                else
+                {
+                    return 0;
+                }
+            }
+        }
```

Comment:
```suggestion
        public static long WorkingSet => Interop.procfs.GetResidentSetSizeById(ProcessId);
```

Reply:
I think it may be the combination of stackalloc and the unsafe struct type?

Reply:
BTW, I came up with what's in the PR after some google searches like
"How to read binary data into C# struct" and the like.
If there's a better or canonical way to do that, I'd appreciate pointers to documentation and examples.

Reply:
In the proposed changes, the unsafe struct would be returned out to consumers.
In the version I have up for review, I've tried to keep the unsafe structs private to that Introp. module.
At the moment I like that strategy better.

I can look into other ways we might avoid building an entire ProcessInfo for the Environment caller,
assuming that was the reason for wanting to expose the unsafe struct to that consumer.

Reply:
If you are on the system, you should try debugging the code to find out why it's not working without extra copies variant. Making multiple copies of data for internal use is not a good idea when it comes to core .NET libraries which every user application is going to depend on. Making extra copy of struct does not make anything "safe".

Try dumping and analyzing data specifically the state of psinfo after fs.ReadExactly line. You may want to change that to ReadAtLeast etc. depending on the finding. I just wrote the compilable code. I'm not on illumos machine right now.

> after some google searches like "How to read binary data into C# struct" and the like.

The google searches and AI results usually don't give the "modern" and efficient way of writing interop code. You should look at interop guidelines links I already shared with you a few times.

Reply:
> If you are on the system, you should try debugging the code to find out why it's not working without extra copies variant. Making multiple copies of data for internal use is not a good idea when it comes to core .NET libraries which every user application is going to depend on. Making extra copy of struct does not make anything "safe".
> 
> Try dumping and analyzing data specifically the state of psinfo after fs.ReadExactly line. You may want to change that to ReadAtLeast etc. depending on the finding. I just wrote the compilable code. I'm not on illumos machine right now.
> 
 
> The google searches and AI results usually don't give the "modern" and efficient way of writing interop code. You should look at interop guidelines links I already shared with you a few times.

I have looked, but I did not see examples using these `unsafe` structs with fixed sized contents
and reading directly into them.

For reasons I haven't yet figured out, when that thing is returned, it looks wrong.
What I read about the stackalloc feature, we might not be able to return that.
(documentation suggest the lifetime ends at function return? not sure)
I'll work on that some more to see if I can get something like that to work.

I'm also struggling with debug output.  Debug.WriteLine calls appear to do nothing.
(Is there a way to turn those on or something?)
Attempts to add Console.WriteLine to the Interop call gives compiler errors indicating
that's not available.

Reply:
Just inline ReadPsInfo in TryGetProcessInfoById and WorkingSet without splitting the method.

Reply:
Just inline ReadPsInfo in TryGetProcessInfoById and WorkingSet without splitting the method. I did it to demonstrate how to deduplicate the functionality. Another option is to split at `Span<byte>` boundary:

```c#
internal static unsafe bool TryGetProcessInfoById(int pid, out ProcessInfo result)
{
    int size = sizeof(psinfo);
    Debug.Assert(size <= 1024, "psinfo struct size exceeds 1024 bytes.");
    Span<byte> buffer = stackalloc byte[size];

    if (!ReadPsInfo(pid, buffer))
    {
        result = default;
        return false;
    }

    ref psinfo psi = ref MemoryMarshal.AsRef<psinfo>(buffer);
....rest of the assignments
}

internal static bool ReadPsInfo(int pid, Span<byte> buffer)
{...
```

Reply:
Just inline ReadPsInfo in TryGetProcessInfoById and WorkingSet without splitting the method. I did it to demonstrate how to deduplicate the functionality. Another option is to split at `Span<byte>` boundary:

```c#
internal static unsafe bool TryGetProcessInfoById(int pid, out ProcessInfo result)
{
    int size = sizeof(psinfo);
    Debug.Assert(size <= 1024, "psinfo struct size exceeds 1024 bytes.");
    Span<byte> buffer = stackalloc byte[size];

    if (!ReadRawPSInfo(pid, buffer))
    {
        result = default;
        return false;
    }

    psinfo* psi = (psinfo*)buffer;
....rest of the assignments
}

internal static bool ReadRawPSInfo(int pid, Span<byte> buffer)
{...
```

Reply:
OK, but I'm confused about how this line is meant to work:
```
    psinfo* psi = (psinfo*)buffer;
```
where buffer is from stackalloc.  I think that means buffer, as well as the thing that we return is no longer "live" after we return, and can be garbage.  Or is there something I'm missing?
Can we allocate a buffer instead here?  I tried, but ran into byte[] vs byte* issues...

Reply:
Here is the refactoring, with latest interop and lang stack https://github.com/dotnet/runtime/commit/6ecc219ec10371c2989d5427f8b17c13e6cf596a. This version doesn't require any `unsafe`-ness (I haven't tested the full build so don't panic if it doesn't work in the first attempt)

Reply:
Here is the refactored version, with latest interop and lang stack https://github.com/dotnet/runtime/commit/6ecc219ec10371c2989d5427f8b17c13e6cf596a. This version doesn't require any `unsafe`-ness (I haven't tested the full build so don't panic if it doesn't work in the first attempt)

Reply:
OK, took your advice and kept the reading stuff in-line, but now using a stackalloc buffer.
This still passes tests.

Reply:
If InlineArray version without unsafe works, that will be a bonus. https://github.com/am11/runtime/commit/1de18084e6fafec869cb8fdc9b9269c7a9b953b3

Reply:
See https://github.com/dotnet/runtime/pull/105403#issuecomment-2259843494, that doesn't require `unsafe` using InlineArray.

Reply:
> I think you could eliminate the `unsafe` from these methods like this:
> 
> ```cs
> psinfo pr = default;
> fs.ReadExactly(MemoryMarshal.AsBytes(new Span<psinfo>(ref pr));
> ```
> 
> The main downside, compared to the current `stackalloc` approach, is that this zero-initializes the memory unnecessarily. If you want to avoid that you could write:
> 
> ```cs
> psinfo pr;
> Unsafe.SkipInit(out pr);
> fs.ReadExactly(MemoryMarshal.AsBytes(new Span<psinfo>(ref pr));
> ```

I was able to replace the stackalloc calls using your suggestions. (pushed to the PR branch)
I tried removing unsafe but then the C# compiler was complaining.

After reading about basically "what does unsafe mean", I'm really not that concerned about the "unsafe" classification for the two functions: `TryGetProcessInfoById` and `TryGetThreadInfoById`.  Those are not doing anything all that complicated, so I'm not sure why "unsafe" should be a big concern.  I'm sure the C# experts will educate me :)

There are things I like about the current approach.  I like that the details of struct psinfo and struct lwpsinfo are kept private to the code that reads them.  I think keeping that detail hidden from consumers is a good thing.  (Just the general, software engineering idea of information hiding, and "narrow" interfaces.)
I also like that those two struct definitions closely resemble the ones the system provides (sys/procfs.h) and would see increasing divergence between those two as contributing to difficulty comparing the two.

This works, and I'd really rather get on to completing features before spending time on performance or perfection.

Reply:
>  I tried removing unsafe but then the C# compiler was complaining.

What was the error message when you tried the suggestion?

>  I also like that those two struct definitions closely resemble the ones the system provides (sys/procfs.h) and would see increasing divergence between those two as contributing to difficulty comparing the two.

You mean syntactically right? Functionally they should be the same. Neither `fixed` keyword nor `InlineArray` resemble C, because C# (like Python, Rust and Go etc.) is not C.

> I'd really rather get on to completing features before spending time on performance or perfection.

This PR is not going to get merged before the branch is opened for 10.0, so you have plenty of time.

> Just the general, software engineering idea of information hiding, and "narrow" interfaces.

That applies to public APIs, not the internal partial classes. In core libraries, we use all kind of quirks to avoid extra copies, so applying the suggestion won't be violating any principles here.

Reply:
Actually, with a look at the GC impact, I like the stackalloc version better (the version just before the latest change).  With the latest version, psinfo and lwpsinfo create garbage on every call. (right?)

With the stackalloc version, the only object created per process or per thread is the result (ProcessInfo, ThreadInfo) which is small.  Yes that implies a little copying, but not enough to worry me.